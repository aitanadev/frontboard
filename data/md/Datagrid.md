# Data grid

`DataGrid` `c-data-grid`

Data grid component allow you to compose any kind of tabular data visualization.

Datagrid supports:

* Sticky columns
* Custom column resize
* Column _drag&drop_ positioning
* Row _drag&drop_
* Bulk operations and row multiselect
* Available **slots** for header cells, regular cells and named cells

```html
<template>
  <Datagrid :cols="cols" :records="records"/>
</template>
<script>
({
 records: [
    {id: '1', name: 'name', data: 10, value: true},
    {id: '2', name: 'name', data: 10, value: true},
    {id: '3', name: 'name', data: 10, value: true},
    {id: '4', name: 'name', data: 10, value: true},
    {id: '5', name: 'name', data: 10, value: true},
    {id: '6', name: 'name', data: 10, value: true},
    {id: '7', name: 'name', data: 10, value: true},
    {id: '8', name: 'name', data: 10, value: true},
    {id: '9', name: 'name', data: 10, value: true},
    {id: '10', name: 'name', data: 10, value: true},
    {id: '11', name: 'name', data: 10, value: true},
    {id: '12', name: 'name', data: 10, value: true},
    {id: '13', name: 'name', data: 10, value: true},
    {id: '14', name: 'name', data: 10, value: true},
    {id: '15', name: 'name', data: 10, value: true},
    {id: '16', name: 'name', data: 10, value: true},
    {id: '17', name: 'name', data: 10, value: true},
    {id: '18', name: 'name end', data: 10, value: true}
  ],
  cols: [
    {key: 'id', col: true, size: 100, sticky: 'left', label: 'Id'},
    {key: 'name', col: true, size: 200, sticky: false, label: 'Name'},
    {key: 'data', col: true, size: 200, sticky: false, label: 'Data'},
    {key: 'value', col: true, size: 200, sticky: false, label: 'Value'},
    {key: 'actions', col: true, size: 100, sticky: 'right', label: 'Actions'}
  ]
})
</script>
```
<!--- section --->
# Injected contents

```html
<template>
  <Datagrid
    :cols="[
      {key: 'id', col: true, size: 80, sticky: 'left', label: 'Id'},
      {key: 'name', col: true, size: 200, sticky: false, label: 'Name'},
      {key: 'data', col: true, size: 150, sticky: false, label: 'Data'},
      {key: 'value', col: true, size: 300, sticky: 'right', label: 'Value'}
    ]"
  >
    <record>
      <cell>1</cell><cell><strong>name 1</strong></cell><cell>😀</cell><cell>true</cell>
    </record>
    <record>
      <cell>2</cell><cell><em>name 2</em></cell><cell>🙂</cell><cell>false</cell>
    </record>
    <record>
      <cell>3</cell><cell>name 3</cell><cell>😁</cell><cell>true</cell>
    </record>
    <record>
      <cell>4</cell><cell>name 4</cell><cell>😃</cell><cell>false</cell>
    </record>
    <record>
      <cell>5</cell><cell>name 5</cell><cell>😄</cell><cell>true</cell>
    </record>
    <record>
      <cell>6</cell><cell>name 6</cell><cell>🙃</cell><cell>false</cell>
    </record>
  </Datagrid>
</template>
```
<!--- section --->
# Injected contents advance

```html
<template>
  <Datagrid
    selectable
    sortable
    :colsPrepend="[
      {key: '$', col: true, size: 90, sticky: 'left', label: ''},
    ]"
    :cols="[
      {key: 'id', col: true, size: 80, sticky: false, label: 'Id'},
      {key: 'name', col: true, size: 200, sticky: false, label: 'Name'},
      {key: 'data', col: true, size: 150, sticky: false, label: 'Data'},
      {key: 'value', col: true, size: 300, sticky: false, label: 'Value'},
      {key: 'action', col: true, size: 120, sticky: 'right', label: 'Actions'}
    ]"
    @selectionChange="window.selection = $event;$forceUpdate()"
  >
    <record>
      <cell>1</cell><cell><strong>name 1</strong></cell><cell>😀</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>2</cell><cell><em>name 2</em></cell><cell>🙂</cell><cell>false</cell><cell>...</cell>
    </record>
    <record>
      <cell>3</cell><cell>name 3</cell><cell>😁</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>4</cell><cell>name 4</cell><cell>😃</cell><cell>false</cell><cell>...</cell>
    </record>
    <record>
      <cell>5</cell><cell>name 5</cell><cell>😄</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>6</cell><cell>name 6</cell><cell>🙃</cell><cell>false</cell><cell>...</cell>
    </record>
    <record>
      <cell>1</cell><cell><strong>name 1</strong></cell><cell>😀</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>2</cell><cell><em>name 2</em></cell><cell>🙂</cell><cell>false</cell><cell>...</cell>
    </record>
    <record>
      <cell>3</cell><cell>name 3</cell><cell>😁</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>4</cell><cell>name 4</cell><cell>😃</cell><cell>false</cell><cell>...</cell>
    </record>
    <record>
      <cell>5</cell><cell>name 5</cell><cell>😄</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>6</cell><cell>name 6</cell><cell>🙃</cell><cell>false</cell><cell>...</cell>
    </record>
    <record>
      <cell>1</cell><cell><strong>name 1</strong></cell><cell>😀</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>2</cell><cell><em>name 2</em></cell><cell>🙂</cell><cell>false</cell><cell>...</cell>
    </record>
    <record>
      <cell>3</cell><cell>name 3</cell><cell>😁</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>4</cell><cell>name 4</cell><cell>😃</cell><cell>false</cell><cell>...</cell>
    </record>
    <record>
      <cell>5</cell><cell>name 5</cell><cell>😄</cell><cell>true</cell><cell>...</cell>
    </record>
    <record>
      <cell>6</cell><cell>name 6</cell><cell>🙃</cell><cell>false</cell><cell>...</cell>
    </record>
  </Datagrid>
  {{ window.selection?.map(record => record.data) }}
</template>
```