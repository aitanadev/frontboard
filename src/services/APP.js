if (!window.$APPCONFIG) throw new Error('Missing $APPCONFIG')

const config = window.$APPCONFIG
const enviroment = config.enviroment
const isDevelopment = enviroment === 'development'
const isStaging = enviroment === 'staging'
const isWebpackServe = process.env.WEBPACK_SERVE
const isJest = process.env.JEST_WORKER_ID
const isOffline = config.offline

const APP = {
  // processenv: process.env,
  config,
  isDevelopment,
  isStaging,
  isWebpackServe,
  isJest,
  isOffline
}

export default APP
