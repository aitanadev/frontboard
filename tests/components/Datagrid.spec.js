import APP from '#services/APP'
import Datagrid from '#components/Datagrid'
import { mount } from '@vue/test-utils'

describe('Datagrid.vue', () => {
  const wrapper = mount(Datagrid, {
    propsData: {
      records: [
        {id: '1', name: 'name', data: 10, value: true},
        {id: '2', name: 'name', data: 10, value: true},
        {id: '3', name: 'name', data: 10, value: true},
        {id: '4', name: 'name', data: 10, value: true},
        {id: '5', name: 'name', data: 10, value: true},
        {id: '6', name: 'name', data: 10, value: true},
        {id: '7', name: 'name', data: 10, value: true},
        {id: '8', name: 'name', data: 10, value: true},
        {id: '9', name: 'name', data: 10, value: true},
        {id: '10', name: 'name', data: 10, value: true},
        {id: '11', name: 'name', data: 10, value: true},
        {id: '12', name: 'name', data: 10, value: true},
        {id: '13', name: 'name', data: 10, value: true},
        {id: '14', name: 'name', data: 10, value: true},
        {id: '15', name: 'name', data: 10, value: true},
        {id: '16', name: 'name', data: 10, value: true},
        {id: '17', name: 'name', data: 10, value: true},
        {id: '18', name: 'name end', data: 10, value: true}
      ],
      cols: [
        {key: 'id', size: 50, sticky: 'left', label: 'Id'},
        {key: 'name', size: 200, sticky: false, label: 'Name'},
        {key: 'data', size: 200, sticky: false, label: 'Data'},
        {key: 'value', size: 90, sticky: 'right', label: 'Value'}
      ]
    }
  })
  beforeAll(async () => {
    await wrapper.vm.$nextTick()
  })
  it('have correct setup', async () => {
    // expect(wrapper.vm.prop).toBe(true)
  })
  it('initial state', async () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it('change', async () => {
    wrapper.vm.cols.pop()
    await wrapper.vm.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
  })
})
