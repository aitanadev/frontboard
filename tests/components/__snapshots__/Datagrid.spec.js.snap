// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datagrid.vue change 1`] = `
<div
  class="c-datagrid"
>
  <table
    class="--scroll-none"
    style="width: 450px;"
  >
    <thead>
      <tr>
        <th
          class="--sticky-left --stickable"
          style="width: 50px; left: 0px;"
        >
          <div>
            Id
          </div>
        </th>
        <th
          class=" --stickable"
          style="width: 200px;"
        >
          <div>
            Name
          </div>
        </th>
        <th
          class=" --stickable"
          style="width: 200px;"
        >
          <div>
            Data
          </div>
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            1
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            2
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            3
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            4
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            5
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            6
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            7
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            8
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            9
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            10
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            11
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            12
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            13
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            14
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            15
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            16
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            17
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            18
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name end
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
      </tr>
    </tbody>
     
    <tfoot>
      <tr>
        <td
          colspan="3"
        />
      </tr>
    </tfoot>
  </table>
</div>
`;

exports[`Datagrid.vue initial state 1`] = `
<div
  class="c-datagrid"
>
  <table
    class="--scroll-none"
    style="width: 540px;"
  >
    <thead>
      <tr>
        <th
          class="--sticky-left --stickable"
          style="width: 50px; left: 0px;"
        >
          <div>
            Id
          </div>
        </th>
        <th
          class=" --stickable"
          style="width: 200px;"
        >
          <div>
            Name
          </div>
        </th>
        <th
          class=" --stickable"
          style="width: 200px;"
        >
          <div>
            Data
          </div>
        </th>
        <th
          class="--sticky-right --stickable"
          style="width: 90px; right: 0px;"
        >
          <div>
            Value
          </div>
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            1
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            2
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            3
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            4
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            5
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            6
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            7
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            8
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            9
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            10
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            11
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            12
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            13
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            14
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            15
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            16
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            17
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="--sticky-left"
          style="left: 0px;"
        >
          <div>
            18
          </div>
        </td>
        <td
          class=""
        >
          <div>
            name end
          </div>
        </td>
        <td
          class=""
        >
          <div>
            10
          </div>
        </td>
        <td
          class="--sticky-right"
          style="right: 0px;"
        >
          <div>
            true
          </div>
        </td>
      </tr>
    </tbody>
     
    <tfoot>
      <tr>
        <td
          colspan="4"
        />
      </tr>
    </tfoot>
  </table>
</div>
`;
